Rob/1.36.0 docs

robfusion opened the pull request:
First iteration of SCIM support documentation

mooreds commented on site/docs/v1/tech/apis/scim/SCIMEnterpriseUser.adoc:
```diff
@@ -0,0 +1,159 @@
+
+== Delete a EnterpriseUser
+
+This API is used to delete a FusionAuth User through a SCIM request. You must specify the Id of the User on the URI.
```
```suggestion
This API is used to hard delete a FusionAuth User through a SCIM request. You must specify the Id of the User on the URI.
```

mooreds commented on site/docs/v1/tech/apis/scim/SCIMEnterpriseUser.adoc:
```diff
@@ -0,0 +1,159 @@
+
+== Update a EnterpriseUser
+
+This API is used to update a new FusionAuth User from a SCIM request. This is a true update meaning that the FusionAuth User will be overwritten with only the data contained in the request. It is NOT a partial update.
```
```suggestion
This API is used to update a new FusionAuth User from a SCIM request. The FusionAuth User will be overwritten by  the data contained in the request. It is not a partial update or a patch. The only exceptions are ...
```

I assume we aren't updating the password, salt, encryptionScheme fields with this? Or are we?

robfusion replied:
It will actually update the password per SCIM specification

mooreds commented on site/docs/v1/tech/apis/scim/SCIMEnterpriseUser.adoc:
```diff
@@ -0,0 +1,159 @@
+
+=== Response
+
+The response for this API contains the User that was just updated in SCIM schema format.
```
```suggestion
The response for this API contains the User that was updated in SCIM schema format.
```

mooreds commented on site/docs/v1/tech/apis/scim/SCIMEnterpriseUser.adoc:
```diff
@@ -0,0 +1,159 @@
+:create_user!:
+:webhook_event!:
+
+For FusionAuth SCIM endpoints, any error responses will be returned in standard SCIM schema. See more details in the link:/docs/v1/tech/apis/scim[SCIM API Overview].
```
```suggestion
For SCIM endpoints, error responses will be returned in standard SCIM schema. See more details in the link:/docs/v1/tech/apis/scim[SCIM API Overview].
```

mooreds replied:
I don't know that there is value in saying `FusionAuth SCIM` instead of SCIM. Are there other kinds of SCIM endpoints we'll be documenting?

mooreds commented on site/docs/v1/tech/apis/scim/SCIMGroup.adoc:
```diff
@@ -0,0 +1,159 @@
+:create_user!:
+:webhook_event!:
+
+For FusionAuth SCIM endpoints, any error responses will be returned in standard SCIM schema. See more details in the link:/docs/v1/tech/apis/scim[SCIM API Overview].
```
Is this worth breaking out to an include? I see it often :)

robfusion replied:
You mean just that one sentence? That seems like maybe overkill, but I can do it if you think it's useful.

mooreds commented on site/docs/v1/tech/apis/scim/SCIMGroup.adoc:
```diff
@@ -0,0 +1,159 @@
+
+== Delete a Group
+
+This API is used to delete a FusionAuth Group through a SCIM request. You must specify the Id of the Group on the URI.
```
```suggestion
This API is used to hard delete a FusionAuth Group through a SCIM request. You must specify the Id of the Group on the URI.
```

mooreds commented on site/docs/v1/tech/apis/scim/SCIMGroup.adoc:
```diff
@@ -0,0 +1,159 @@
+
+== Update a Group
+
+This API is used to update a new FusionAuth Group from a SCIM request. This is a true update meaning that the FusionAuth Group will be overwritten with only the data contained in the request. It is NOT a partial update.
```
```suggestion
This API is used to update a new FusionAuth Group from a SCIM request. The FusionAuth Group will be overwritten with only the data contained in the request. It is not a partial update or patch.
```

mooreds commented on site/docs/v1/tech/apis/scim/SCIMGroup.adoc:
```diff
@@ -0,0 +1,159 @@
+
+=== Response
+
+The response for this API contains the Group that was just updated in SCIM schema format.
```
```suggestion
The response for this API contains the Group that was updated in SCIM schema format.
```

mooreds commented on site/docs/v1/tech/apis/scim/SCIMServiceProvider.adoc:
```diff
@@ -0,0 +1,17 @@
+--
+[method]#GET# [uri]#/api/scim/resources/v2/ResourceTypes#
+
+Documentation in progress...
```
```suggestion
```

Just leave it blank. You can also ship it but don't include a link anywhere if you must.

mooreds commented on site/docs/v1/tech/apis/scim/SCIMServiceProvider.adoc:
```diff
@@ -0,0 +1,17 @@
+---
+layout: doc
+title: SCIM Service Provider API Overview
+description: APIs for dsicovering configuration information about FusionAuth SCIM Service Provider using your SCIM Client
```
```suggestion
description: APIs for discovering configuration information about FusionAuth SCIM Service Provider using your SCIM Client
```

mooreds commented on site/docs/v1/tech/apis/scim/SCIMUser.adoc:
```diff
@@ -0,0 +1,159 @@
+
+=== Response
+
+The response for this API contains the User that was just created in SCIM schema format.
```
```suggestion
The response for this API contains the User that was created in SCIM schema format.
```

mooreds commented on site/docs/v1/tech/apis/scim/SCIMUser.adoc:
```diff
@@ -0,0 +1,159 @@
+
+== Delete a User
+
+This API is used to delete a FusionAuth User through a SCIM request. You must specify the Id of the User on the URI.
```
```suggestion
This API is used to hard delete a FusionAuth User through a SCIM request. You must specify the Id of the User on the URI.
```

mooreds commented on site/docs/v1/tech/apis/scim/SCIMUser.adoc:
```diff
@@ -0,0 +1,159 @@
+
+== Update a User
+
+This API is used to update a new FusionAuth User from a SCIM request. This is a true update meaning that the FusionAuth User will be overwritten with only the data contained in the request. It is NOT a partial update.
```
```suggestion
This API is used to update a new FusionAuth User from a SCIM request. The FusionAuth User will be overwritten with only the data contained in the request. It is not a partial update or a patch.
```

mooreds replied:
What about password, etc?

mooreds commented on site/docs/v1/tech/apis/scim/SCIMUser.adoc:
```diff
@@ -0,0 +1,159 @@
+
+=== Response
+
+The response for this API contains the User that was just updated in SCIM schema format.
```
```suggestion
The response for this API contains the User that was updated in SCIM schema format.
```

mooreds commented on site/docs/v1/tech/apis/scim/_scim-enterpriseuser-request-body.adoc:
```diff
@@ -0,0 +1,72 @@
+Because SCIM schemas are completely customizable, There are no required fields for the incoming request other than the common fields that all SCIM requests must include to match the specification. The only requirement about the user data is that in the incoming request lambda, you must specify either the **email** or the **username** or both for the FusionAuth User object. Either of these values  ...
```
```suggestion
Because SCIM schemas are completely customizable, there are no required fields for the incoming request other than the common fields that all SCIM requests must include to match the specification.
The only requirement about the user data is that in the incoming request lambda, you must specify either the **email**, the **username**, or both for the FusionAuth User object.
Either of these values may uniquely identify the User and can be used during a login.
For all other fiends, it is completely up to your lambda to map incoming user data to the appropriate FusionAuth User fields.
```

mooreds replied:
If you have multiple lines of text with no newlines between them, Asciidoc collapses them to a paragraph. Multiple lines are nice for diffs.

mooreds commented on site/docs/v1/tech/apis/scim/_scim-enterpriseuser-request-body.adoc:
```diff
@@ -0,0 +1,72 @@
+Because SCIM schemas are completely customizable, There are no required fields for the incoming request other than the common fields that all SCIM requests must include to match the specification. The only requirement about the user data is that in the incoming request lambda, you must specify either the **email** or the **username** or both for the FusionAuth User object. Either of these values  ...
+
+The following is an example of a typical SCIM User create request body your SCIM client might send. However, your incoming request lambda could use these values in completely different ways to populate the FusionAuth User object if you choose.
```
I'm confused about the "however" sentence. Seems like it belongs later in the process because here we are talking about the input before the mapping, right?

mooreds commented on site/docs/v1/tech/apis/scim/_scim-enterpriseuser-request-body.adoc:
```diff
@@ -0,0 +1,72 @@
+
+This is taken from the SCIM Schema RFC describing a SCIM User schema. For the full specification you can find the RFC here:
+
+- https://datatracker.ietf.org/doc/html/rfc7643#page-19
```
You could just put a link rather than have a sep bullet point.

mooreds commented on site/docs/v1/tech/apis/scim/_scim-enterpriseuser-request-body.adoc:
```diff
@@ -0,0 +1,72 @@
+
+The following is an example of a typical SCIM User create request body your SCIM client might send. However, your incoming request lambda could use these values in completely different ways to populate the FusionAuth User object if you choose.
+
+This is taken from the SCIM Schema RFC describing a SCIM User schema. For the full specification you can find the RFC here:
```
```suggestion
This is taken from the SCIM Schema RFC describing a SCIM User schema. https://datatracker.ietf.org/doc/html/rfc7643#page-19[Read the full specification here].
```

mooreds replied:
Also, point to a section, not a page

mooreds commented on site/docs/v1/tech/apis/scim/_scim-enterpriseuser-request-body.adoc:
```diff
@@ -0,0 +1,72 @@
+The externalId is a unique identifier created by the client that identifies the resource in the SCIM client's environment. It could be used as the FusionAuth resource id as well, but that is not a requirement of the SCIM specification. That would need to be done by the defined incoming lambda if desired.
+
+[field]#name.formatted# [type]#[String]# [optional]#Optional#::
+his would typically be the value used to populate the FusionAuth User.fullName by the incoming request lambda.
```
```suggestion
This would typically be the value used to populate the FusionAuth User.fullName by the incoming request lambda.
```

mooreds replied:
The `T` is missing for many of these fields.

mooreds commented on site/docs/v1/tech/apis/scim/_scim-enterpriseuser-request-body.adoc:
```diff
@@ -0,0 +1,72 @@
+The state of the SCIM Resource.
+
+[field]#schemas# [type]#[Array]# [required]#Required#::
+The String identifier of the SCIM Resource schemas and extensions provided in the request. This should match supported schemas and extensions for the provisioning server.
```
Do we only handle certain schemas?

robfusion replied:
Yes and no. Technically they can handle whatever they want in the Lambda as long as it maps the incoming schema values to FusionAuth concepts like User and Group. They are free the advertise that they support whatever schemas and customizations they want, but ultimately they are responsible for converting the requests and responses in their lambdas.

mooreds commented on site/docs/v1/tech/apis/scim/_scim-enterpriseuser-request-body.adoc:
```diff
@@ -0,0 +1,72 @@
+This would typically be the value used to populate The FusionAuth User.password by the incoming request lambda.
+
+[field]#phoneNumbers# [type]#[Array]# [optional]#Optional#::
+SCIM supports multiple phone numbers while FusionAuth only allows a single phone number for a user. The incoming request lambda would need to determine which one to use if multiple are specified. The most common strategy would be to just use the one where primary is true.
```
`FusionAuth only allows a single phone number for a user` -> `FusionAuth only has a single field, mobilePhone. Additional phone numbers can be stored in the data field`

mooreds commented on site/docs/v1/tech/apis/scim/_scim-enterpriseuser-request-body.adoc:
```diff
@@ -0,0 +1,72 @@
+This would typically be used to populate the FusionAuth User.mobilePhone by the incoming request lambda.
+
+[field]#phoneNumbers[x].type# [type]#[String]# [optional]#Optional#::
+This might be used to alternatively choose a value to decide which phone number to use to populate the FusionAuth User.mobilePhone by the incoming request lambda.
```
```suggestion
The type of the phone number (work, home, mobile). Rather than use [field]#phoneNumbers[x].primary#, you could select a different number as the FusionAuth User.mobilePhone.
```

mooreds commented on site/docs/v1/tech/apis/scim/_scim-enterpriseuser-request-body.adoc:
```diff
@@ -0,0 +1,72 @@
+
+- https://datatracker.ietf.org/doc/html/rfc7643#page-19
+
+==== Request Body
```
You kind of make this serve two purposes:

* a sample request body, from the spec
* suggested mappings to the FusionAuth user object

I'd suggest doing just the first, and then building a table that offers a typical mapping. This could be a different section of the API docs.

mooreds commented on site/docs/v1/tech/apis/scim/_scim-enterpriseuser-response-body.adoc:
```diff
@@ -0,0 +1,69 @@
+Because SCIM schemas are completely customizable, The SCIM response returned is generated from the result of SCIM User object populated by your outgoing response lambda.
+
+The following is an example of a typical SCIM User response body that your SCIM client might expect. However, your outgoing response lambda could generate a completely different SCIM schema is desired.
```
is it a "typical SCIM User response body that your SCIM client might expect" or one that our default lambda implementation returns, or both? I think the latter is more important to mention.

mooreds commented on site/docs/v1/tech/apis/scim/_scim-enterpriseuser-response-body.adoc:
```diff
@@ -0,0 +1,69 @@
+
+This is taken from the SCIM Schema RFC describing a SCIM User schema. For the full specification you can find the RFC here:
+
+- https://datatracker.ietf.org/doc/html/rfc7643#page-19
```
Same as above, you can move the url inline.

mooreds replied:
And point to a section, not a page

robfusion replied:
done

mooreds commented on site/docs/v1/tech/apis/scim/_scim-enterpriseuser-response-body.adoc:
```diff
@@ -0,0 +1,69 @@
+The externalId is a unique identifier created by the client that identifies the resource in the SCIM client's environment. It could be used as the FusionAuth resource id as well, but that is not a requirement of the SCIM specification. That would need to be done by the defined incoming lambda if desired.
+
+[field]#name.formatted# [type]#[String]#::
+his would typically be the value used to populate the FusionAuth User.fullName by the incoming request lambda.
```
Is the incoming request lambda part of this at all?

mooreds commented on site/docs/v1/tech/apis/scim/_scim-enterpriseuser-response-body.adoc:
```diff
@@ -0,0 +1,69 @@
+- https://datatracker.ietf.org/doc/html/rfc7643#page-19
+
+==== Response Body
+[.api]
```
You kind of make this serve two purposes:

* suggested mappings from the FusionAuth user object
* a sample response body, from the spec perhaps?

I'd suggest doing just the second, and then building a table that offers a typical mapping. This could be a different section of the API docs.

robfusion replied:
Not sure I understand your suggestion exactly. You want me to just remove all the [field] descriptions and just have a plain JSON response body with nothing else?

robfusion replied:
Well, if we remove the advice on mapping then it's really just a copy of the spec. Maybe we should jsut link to the spec and not provide an example request and response JSON?

mooreds commented on site/docs/v1/tech/apis/scim/_scim-group-request-body.adoc:
```diff
@@ -0,0 +1,38 @@
+This is taken from the SCIM Schema RFC describing a SCIM User schema. For the full specification you can find the RFC here:
+
+- https://datatracker.ietf.org/doc/html/rfc7643#section-4.2
+
```
Same comments about splitting up mapping and field definitions.

mooreds commented on site/docs/v1/tech/apis/scim/_scim-group-request-body.adoc:
```diff
@@ -0,0 +1,38 @@
+
+This is taken from the SCIM Schema RFC describing a SCIM User schema. For the full specification you can find the RFC here:
+
+- https://datatracker.ietf.org/doc/html/rfc7643#section-4.2
```
Would have this inline above.

robfusion replied:
done

mooreds commented on site/docs/v1/tech/apis/scim/_scim-user-request-body.adoc:
```diff
@@ -0,0 +1,72 @@
+
+This is taken from the SCIM Schema RFC describing a SCIM User schema. For the full specification you can find the RFC here:
+
+- https://datatracker.ietf.org/doc/html/rfc7643#page-19
```
Would have this inline. Also, point to a section, not a page

mooreds commented on site/docs/v1/tech/apis/scim/_scim-user-response-body.adoc:
```diff
@@ -0,0 +1,72 @@
+
+This is taken from the SCIM Schema RFC describing a SCIM User schema. For the full specification you can find the RFC here:
+
+- https://datatracker.ietf.org/doc/html/rfc7643#page-19
```
Would have this inline, above. Also point to a section, not a page.

robfusion replied:
done

mooreds commented on site/docs/v1/tech/apis/scim/_scim-group-response-body.adoc:
```diff
@@ -0,0 +1,53 @@
+Because SCIM schemas are completely customizable, The SCIM response returned is generated from the result of SCIM Group object populated by your outgoing response lambda.
+
```
Same comment about the mapping and the field definitions.

robfusion replied:
done

mooreds commented on site/docs/v1/tech/apis/scim/_scim-group-response-body.adoc:
```diff
@@ -0,0 +1,53 @@
+
+This is taken from the SCIM Schema RFC describing a SCIM Group schema. For the full specification you can find the RFC here:
+
+- https://datatracker.ietf.org/doc/html/rfc7643#section-4.2
```
Move inline.

robfusion replied:
done

mooreds commented on site/docs/v1/tech/apis/scim/_scim-user-request-body.adoc:
```diff
@@ -0,0 +1,72 @@
+Because SCIM schemas are completely customizable, there are no required fields for the incoming request other than the common fields that all SCIM requests must include to match the specification. The only requirement about the user data is that in the incoming request lambda, you must specify either the **email** or the **username** or both for the FusionAuth User object. Either of these values  ...
```
```
Because SCIM schemas are completely customizable, there are no required fields for the incoming request other than the common fields that all SCIM requests must include to match the specification.
The only requirement about the user data is that in the incoming request lambda, you must specify either the **email**, the **username**, or both for the FusionAuth User object.
Either of these values may uniquely identify the User and can be used during a login.
For all other fiends, it is completely up to your lambda to map incoming user data to the appropriate FusionAuth User fields.
```

mooreds replied:
Also would break this text out to an include file, as it appears to be shared

mooreds commented on site/docs/v1/tech/apis/scim/_scim-user-request-body.adoc:
```diff
@@ -0,0 +1,72 @@
+[.api]
+[field]#active# [type]#[Boolean]# [required]#Required#::
+The state of the SCIM Resource.
+
```
How is this different than the enterprise user request? Can you collapse them?

robfusion replied:
EnterpriseUser is really an extension of User. I can use includes there.

robfusion replied:
I was missing some of the field descriptions before. SHould have descriptions for everything now.

mooreds commented on site/docs/v1/tech/apis/scim/_scim-user-response-body.adoc:
```diff
@@ -0,0 +1,72 @@
+==== Response Body
+[.api]
+[field]#active# [type]#[Boolean]#::
+The state of the SCIM Resource. Typically just mapped from the FusionAuth User.active property.
```
How is this different than the enterprise user response? Can you collapse?

mooreds replied:
Or create includes to share?

robfusion replied:
EnterpriseUser is really an extension of User. I can use includes there.

robfusion replied:
Not sure of the best way to handle this since EnterpriseUser is really just a schema extension so I included the User schema in the request anyway.

robfusion replied:
refactor response body into includes as well

robfusion replied:
Should be done now.

mooreds commented on site/docs/v1/tech/apis/scim/index.adoc:
```diff
@@ -0,0 +1,43 @@
+
+FusionAuth also supports the SCIM service provider discovery endpoints per the specification
+
+* link:/docs/v1/tech/apis/scim/SCIMServiceProvider[SCIM Service Provider]
```
Would omit this until it is documented.

robfusion replied:
done

mooreds commented on site/docs/v1/tech/apis/scim/index.adoc:
```diff
@@ -0,0 +1,43 @@
+
+All error responses from From FusionAuth SCIM API endpoints will be returned in SCIM compliant schem as defined here.
+
+- https://datatracker.ietf.org/doc/html/rfc7644#section-3.12
```
Move up and link inline

robfusion replied:
done

mooreds reviewed (changes requested):
* do we ship with default scim entities? Should we? If we do, we should reference them. I didn't see them in the sandbox, though. If not, we should document how to create one (what perms do they need?)

* We need a 5th type of authentication, client credentials. Need to pick a new icon and update /docs/v1/tech/apis/authentication

* what is difference between user and enterpriseuser? Or will a scim person know this? I'd add that to the core concepts.
* separate mapping and field definitions for req/res (also added that in comments).
* see if you can leverage includes for enterpriseuser and user req/res. lots of overlap.
* run all docs through google docs and the spelling and grammar check
* If you don't have a doc done, just don't like to it.

robfusion commented on site/docs/v1/tech/core-concepts/scim.adoc:
```diff
@@ -0,0 +1,73 @@
+
+4. FusionAuth will attempt to create the FA User using the returned User data from the incoming request lambda.
++
+5. Upon success, FusionAuth will call the outgoing response lambda passing the newly created FusionAuth User and the SCIMResponse. The outgoing lambda is then responsible for converting the FusionAuth User properties to the appropriate SCIM Response data.
```
Yeah, I think representation is fine. Success = FusionAuth user was created.

robfusion commented on site/docs/v1/tech/apis/scim/index.adoc:
```diff
@@ -0,0 +1,43 @@
+
+== Authentication
+
+In order to use the authenticated FusionAuth SCIM API endpoints, you must create a SCIM client entity and execute the client credentials authorization workflow.
```
done

robfusion commented:
Addressing your comments:

- do we ship with default scim entities? Should we? If we do, we should reference them. I didn't see them in the sandbox, though. If not, we should document how to create one (what perms do they need?)

Yes, still working on docs for those, then I'll link to them.


- We need a 5th type of authentication, client credentials. Need to pick a new icon and update /docs/v1/tech/apis/authentication

pick an icon from where exactly?

- what is difference between user and enterpriseuser? Or will a scim person know this? I'd add that to the core concepts.

EnterpriseUser is a Schema extension of User defined in the SCIM RFC. I added links for User, EnterpriseUser, and Group to the overview in core concepts.

- separate mapping and field definitions for req/res (also added that in comments).

I combined the common fields into include files

- see if you can leverage includes for enterpriseuser and user req/res. lots of overlap.

Did that.

- run all docs through google docs and the spelling and grammar check

Working on that.

- If you don't have a doc done, just don't like to it.

I removed the links to the Service Provider Config docs that aren't complete.

robfusion commented:
Ran everything through Google docs spelling and grammar

robotdan commented on site/docs/src/json/scim/scim-extended-error-response.json:
```diff
@@ -0,0 +1,24 @@
+    }
+  },
+  "schemas": [
+    "io.fusionauth.errors",
```
Isn't this `urn:ietf:params:scim:schemas:extension:fusionauth:2.0:Error`.

robfusion replied:
done

robotdan commented on site/docs/src/json/scim/scim-extended-error-response.json:
```diff
@@ -0,0 +1,24 @@
+{
+  "detail": "Request failed, see errors for additional details.",
+  "io.fusionauth.errors": {
```
`urn:ietf:params:scim:schemas:extension:fusionauth:2.0:Error` and other places in this PR

robfusion replied:
Oh, I forgot we updated that. I'll fix it.

robfusion replied:
done

robotdan commented on site/docs/v1/tech/apis/scim/index.adoc:
```diff
@@ -0,0 +1,37 @@
+---
```
We'll want to add a version flag somewhere on this page to indicate which version we added SCIM Server Support.

robfusion replied:
done

robotdan commented on site/docs/v1/tech/apis/scim/index.adoc:
```diff
@@ -0,0 +1,37 @@
+---
+layout: doc
+title: SCIM API Overview
```
SCIM Server API?

robfusion replied:
done

robotdan commented on site/docs/v1/tech/core-concepts/scim.adoc:
```diff
@@ -0,0 +1,73 @@
+---
+layout: doc
+title: Search
```
Update title and description?

robfusion replied:
done

robotdan commented on site/docs/v1/tech/core-concepts/scim.adoc:
```diff
@@ -0,0 +1,73 @@
+description: Search options within FusionAuth
+---
+
+:sectnumlevels: 0
```
Should probably add  version flag somewhere here?

robfusion replied:
done

robfusion commented on site/docs/v1/tech/core-concepts/tenants.adoc:
```diff
@@ -436,6 +437,14 @@ image::core-concepts/tenant-configuration-webhooks-settings.png[Tenant Configura
 include::docs/v1/tech/events-webhooks/_tenant-webhooks-table.adoc[]
 
+==== SCIM
+
```
Yeah, was planning to add the screenshots etc. later. Just trying to get a first rev of SCIM stuff in instead of trying to do everything in one giant PR. It's already a lot.

mooreds replied:
okay, sounds good. You can always add 

// TODO add screenshots 

and asciidoctor will omit the comment.

robfusion replied:
Added // TODO

robfusion commented on site/docs/v1/tech/core-concepts/scim.adoc:
```diff
@@ -0,0 +1,73 @@
++
++
+- You need to define your incoming request and outgoing response lambdas for each the supported SCIM resource types (User, Enterprise User, Group)
++
```
done

robfusion commented on site/docs/v1/tech/core-concepts/scim.adoc:
```diff
@@ -0,0 +1,73 @@
++
++
+- You will need to verify that the default SCIM schemas provided match your desired SCIM schema for each of the SCIM resource types.
++
```
done

mooreds commented on site/docs/v1/tech/core-concepts/tenants.adoc:
```diff
@@ -436,6 +437,14 @@ image::core-concepts/tenant-configuration-webhooks-settings.png[Tenant Configura
+
+Full documentation on link:/docs/v1/tech/core-concepts/scim/[SCIM] can be found link:/docs/v1/tech/core-concepts/scim/[here].
+
+The SCIM server configuration if you wish for FusionAuth to handle incoming SCIM client provisioning requests
```
```suggestion
The SCIM server configuration to enable incoming SCIM client provisioning requests.
```

robfusion commented on site/docs/v1/tech/apis/scim/SCIMEnterpriseUser.adoc:
```diff
@@ -0,0 +1,159 @@
+
+== Overview
+
+This page contains all of the APIs for managing Users through SCIM EnterpriseUser requests.
```
Yes, it's one of the core schemas defined in the RFC. There is a link to the schemas RFC in the overview as well. I could add another link to it here.

robfusion commented on site/docs/src/json/scim/user-create-response.json:
```diff
@@ -0,0 +1,31 @@
+  "emails" : [ {
+    "primary" : true,
+    "type" : "work",
+    "value" : "tester123@fusionauth.io"
```
Should be the same. Fixing it.

robfusion commented on site/docs/v1/tech/apis/scim/SCIMUser.adoc:
```diff
@@ -0,0 +1,159 @@
+
+== Create a User
+
+This API is used to create a new FusionAuth User from a SCIM request.
```
Different Lambdas would be called to process the incoming and outgoing requests so the dev would be able to retrieve data specific to one schema or the other and choose where to store it in the FusionAuth User. For example, they may want to store a manager's id that could be included in the EnterpriseUser schema into a user.data property since FusionAuth doesn't have that concept. The User schema would not have that data per the the spec so it's lambda would not try to store it.

In practice, the two lambdas will probably be pretty much identical, but this allows for the most flexibility if they want to support both the EnterpriseUser and User SCIM core schemas.

robfusion commented on site/docs/v1/tech/apis/scim/_scim-group-response-body.adoc:
```diff
@@ -0,0 +1,41 @@
+include::/docs/v1/tech/apis/scim/_scim-common-body.adoc[]
+
+[field]#meta.resourceType# [type]#[String]#::
+The SCIM resource type defined by this schema.
```
It's common in all the responses. Missed adding it to the other two. Fixing it.

robfusion replied:
done

robfusion commented on site/docs/v1/tech/apis/scim/_scim-group-request-body.adoc:
```diff
@@ -0,0 +1,29 @@
+[field]#member.displayName# [type]#[String]#::
+
+[field]#meta.$ref# [type]#[String]#::
+The URI that can be used to retrieve data about the Group member.
```
According to the spec the group member is any SCIM resource (User, EnterpriseUser, Group). In FusionAuth it would always be a User that is a member of the group for now at least because I don't believe we support sub groups.

That was a typo in the field though. It should be members[x].$ref. Fixing it.

robfusion replied:
typo fixed.

robfusion commented on site/docs/v1/tech/apis/scim/_scim-enterpriseuser-response-body.adoc:
```diff
@@ -0,0 +1,45 @@
+Identifies the name of a department.
+
+[field]#urn:ietf:params:scim:schemas:extension:enterprise:2.0:User.manager# [type]#[String]# [optional]#Optional#::
+The user's manager.  A complex type that optionally allows service providers to represent organizational hierarchy by referencing the "id" attribute of another User.
```
Well, since we don't really have the concept of "manager" in FA, it really is whatever the user wants it to be.

robfusion commented on site/_layouts/doc.liquid:
```diff
@@ -179,6 +180,15 @@
+                      <a href="#" class="sub-menu">SCIM <i class="fal fa-chevron-{% if page.url contains "/apis/scim" %}up{% else %}down{% endif %} fa-fw"></i></a>
+                      <ul>
+                        <li {% if page.url == "/docs/v1/tech/apis/scim/" %}class="active"{% endif %}><a href="/docs/v1/tech/apis/scim/">Overview</a></li>
+                        <li {% if page.url == "/docs/v1/tech/apis/scim/SCIMGroup" %}class="active"{% endif %}><a href="/docs/v1/tech/apis/scim/SCIMGroup">SCIM Group</a></li>
```
done

robfusion commented on site/docs/v1/tech/core-concepts/scim.adoc:
```diff
@@ -0,0 +1,66 @@
+----
++
+2. FusionAuth will authenticate the incoming request to ensure the request is from a known SCIM client. Each SCIM client and server instance is represented as an Entity and will authenticate using a link:/docs/v1/tech/apis/authentication#client-credentials[Client Credentials Grant].
+* An example link:/docs/v1/tech/oauth/#example-client-credentials-grant[client credentials grant using Entities].
```
done

robfusion commented on site/docs/v1/tech/core-concepts/scim.adoc:
```diff
@@ -0,0 +1,66 @@
+[NOTE.since]
+
+====
+This API has been available since 1.36.0
```
done

robfusion commented on site/docs/v1/tech/apis/scim/_scim-group-response-body.adoc:
```diff
@@ -0,0 +1,41 @@
+[field]#members# [type]#[Array]#::
+This list of members included in this Group.
+
+[field]#member.displayName# [type]#[String]#::
```
done

robfusion commented on site/docs/v1/tech/apis/scim/_scim-group-request-body.adoc:
```diff
@@ -0,0 +1,29 @@
+The URI that can be used to retrieve data about the Group member.
+
+[field]#meta.value# [type]#[String]#::
+The unique id of the Group member.
```
done

robfusion commented on site/docs/v1/tech/apis/scim/_scim-group-request-body.adoc:
```diff
@@ -0,0 +1,29 @@
+[field]#members# [type]#[Array]#::
+This list of members included in this Group.
+
+[field]#member.displayName# [type]#[String]#::
```
done

robfusion commented on site/docs/v1/tech/apis/scim/_scim-group-request-body.adoc:
```diff
@@ -0,0 +1,29 @@
+[.api]
+include::/docs/v1/tech/apis/scim/_scim-common-body.adoc[]
+
+[field]#displayName# [type]#[String]#::
```
done

robfusion commented on site/docs/v1/tech/apis/scim/_scim-group-request-body.adoc:
```diff
@@ -0,0 +1,29 @@
+Because SCIM schemas are completely customizable, there are no required fields for the incoming request other than the common fields that all SCIM requests must include to match the specification.
+
+The following is an example of a typical SCIM Group create request body a SCIM client might send. However, your incoming request lambda could use these values in completely different ways to populate the FusionAuth User object if you choose.
```
done

robfusion commented on site/docs/v1/tech/apis/scim/_scim-enterpriseuser-request-body.adoc:
```diff
@@ -0,0 +1,47 @@
+The user's manager.  A complex type that optionally allows service providers to represent organizational hierarchy by referencing the "id" attribute of another User.
+
+[field]#urn:ietf:params:scim:schemas:extension:enterprise:2.0:User.manager.value# [type]#[String]# [optional]#Optional#::
+The "id" of the SCIM resource representing the user's manager
```
We don't really have the concept of manager so it's up to the dev what they want it to be really, but most likely it would be the FA User Id of the "manager"

robfusion commented on site/docs/v1/tech/apis/scim/_scim-enterpriseuser-request-body.adoc:
```diff
@@ -0,0 +1,47 @@
+Identifies the name of a department.
+
+[field]#urn:ietf:params:scim:schemas:extension:enterprise:2.0:User.manager# [type]#[String]# [optional]#Optional#::
+The user's manager.  A complex type that optionally allows service providers to represent organizational hierarchy by referencing the "id" attribute of another User.
```
FusionAuth is the service provider, but we don't have the concept of hierarchies like that so it would be up to the dev to figure out what they want to do with that data.

robfusion commented on site/docs/v1/tech/apis/scim/_scim-enterpriseuser-request-body.adoc:
```diff
@@ -0,0 +1,47 @@
+This is the root property of the EnterpriseUser schema extension. It should match the identifier included in the schemas property.
+
+[field]#urn:ietf:params:scim:schemas:extension:enterprise:2.0:User.employeeNumber# [type]#[String]# [optional]#Optional#::
+A string identifier, typically numeric or alphanumeric, assigned to a person, typically based on order of hire or association with an organization.
```
Yes, it's just an extension of the SCIM User schema per their spec.

robfusion commented on site/docs/src/json/scim/group-create-response.json:
```diff
@@ -0,0 +1,19 @@
+  "displayName": "Sales Reps",
+  "members": [
+    {
+      "displayName": "Dwight Schrute",
```
Just went with John Doe

robfusion commented on site/docs/v1/tech/core-concepts/scim.adoc:
```diff
@@ -0,0 +1,80 @@
+
+== FusionAuth SCIM Support
+
+Any SCIM client can make requests to FusionAuth using the SCIM specification for defining Users and Groups. In this scenario FusionAuth acts as the provisioning server and can respond with SCIM compliant responses.
```
done

robfusion commented on site/docs/v1/tech/core-concepts/scim.adoc:
```diff
@@ -0,0 +1,66 @@
+4. FusionAuth will attempt to create the FusionAuth User using the mapped object from the incoming request lambda.
++
+5. Upon successful creation of the User, FusionAuth will call the outgoing response lambda, passing the newly created FusionAuth User and the SCIM response. The outgoing lambda is responsible for mapping the FusionAuth User properties to the appropriate SCIM representation.
+
```
done

robotdan commented on site/docs/src/json/scim/group-create-request.json:
```diff
@@ -0,0 +1,12 @@
+  "members": [
+    {
+      "displayName": "John Doe",
+      "$ref": "https://fusionauth.io/api/scim/v2/Users/902c246b-6245-4190-8e05-00816be7344a",
```
I wouldn't use our actual URL here of `https://fusionauth.io/` this is a production URL. Can we use something else? We use `login.piedpiper.com` quite a bit in docs, so that is an option.
